eth/catalyst/simulated_beacon_api.go

```diff
@@ -18,44 +18,89 @@ package catalyst
 
 import (
 	"context"
-	"time"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/types"
-	"github.com/ethereum/go-ethereum/log"
 )
 
-type api struct {
-	sim *SimulatedBeacon
+// simulatedBeaconAPI provides a RPC API for SimulatedBeacon.
+type simulatedBeaconAPI struct {
+	sim      *SimulatedBeacon
+	doCommit chan struct{}
 }
 
-func (a *api) loop() {
+// newSimulatedBeaconAPI returns an instance of simulatedBeaconAPI with a
+// buffered commit channel. If period is zero, it starts a goroutine to handle
+// new tx events.
+func newSimulatedBeaconAPI(sim *SimulatedBeacon) *simulatedBeaconAPI {
+	api := &simulatedBeaconAPI{sim: sim, doCommit: make(chan struct{}, 1)}
+	if sim.period == 0 {
+		// mine on demand if period is set to 0
+		go api.loop()
+	}
+	return api
+}
+
+// loop is the main loop for the API when it's running in period = 0 mode. It
+// ensures that block production is triggered as soon as a new withdrawal or
+// transaction is received.
+func (a *simulatedBeaconAPI) loop() {
 	var (
-		newTxs = make(chan core.NewTxsEvent)
-		sub    = a.sim.eth.TxPool().SubscribeTransactions(newTxs, true)
+		newTxs    = make(chan core.NewTxsEvent)
+		newWxs    = make(chan newWithdrawalsEvent)
+		newTxsSub = a.sim.eth.TxPool().SubscribeTransactions(newTxs, true)
+		newWxsSub = a.sim.withdrawals.subscribe(newWxs)
 	)
-	defer sub.Unsubscribe()
+	defer newTxsSub.Unsubscribe()
+	defer newWxsSub.Unsubscribe()
+
+	go a.worker()
 
 	for {
 		select {
 		case <-a.sim.shutdownCh:
 			return
-		case w := <-a.sim.withdrawals.pending:
-			withdrawals := append(a.sim.withdrawals.gatherPending(9), w)
-			if err := a.sim.sealBlock(withdrawals, uint64(time.Now().Unix())); err != nil {
-				log.Warn("Error performing sealing work", "err", err)
-			}
+		case <-newWxs:
+			a.commit()
 		case <-newTxs:
+			a.commit()
+		}
+	}
+}
+
+// commit is a non-blocking method to initate Commit() on the simulator.
+func (a *simulatedBeaconAPI) commit() {
+	select {
+	case a.doCommit <- struct{}{}:
+	default:
+	}
+}
+
+// worker runs in the background and signals to the simulator when to commit
+// based on messages over doCommit.
+func (a *simulatedBeaconAPI) worker() {
```

Comment:
I have a feeling that both `commit()` and `worker()` doesn't make a lot of sense, other than in the context of `loop()`. And that _maybe_, they should just be closures inside `loop()`? OTOH maybe that just makes things convoluted. 
I'll let you decide

Reply:
I pushed this change. IMO it makes more sense to keep it together, also makes it possible to range over the `doCommit`, since we "own it" inside the `loop`, and are at liberty to close it on exit

Reply:
it looks good inside the loop 👍